Add output

AurelienRenou opened the pull request:
#10 
Add a dict which contain duration, norm of diff vector, number of iteration, ...

AurelienRenou committed:
add output from scipy

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -219,3 +234,21 @@ def animate(self):
         b.load_movement(self.q)
         b.exec()
+
+    def get_sol(self):
+        for i, sol in enumerate(self.sol):
+            for j, value in enumerate(sol.fun):
+                self.residuals_xyz[i][j] = value
+            self.nb_iteration_diff.append(sol.nfev)
+            self.nb_iteration_jac.append(sol.njev)
+
+        for j, value in enumerate(self.residuals_xyz):
+            for h in range(0, 3 * self.nb_markers, 3):
+                self.residuals[j][int(h / 3)] = np.linalg.norm(value[h : h + 3])
+
+        self.output = dict(
+            time=self.duration,
+            residuals=self.residuals,
+            nb_iteration_diff=self.nb_iteration_diff,
+            nb_iteration_jac=self.nb_iteration_jac,
+        )
```
All the variables stored in output should not be attributes of the method.

replace   `self.residuals_xyz[i][j]` by  `residuals_xyz[i][j]`
Same for every other variables.

You should iterate on the `self.nb_frame` instead of the `self.sol`
Try to find a better one for `sol.fun`

To be more generic, you can create a new attribute `self.n_dim` related to the dimension of markers, we could have 2D motion capture so it would be:
`self.nb_dims * self.nb_markers` instead of `*3`

Anyway, your loop needs to be clarified.
Please start your outer loop with the number of markers.
There should be only one loop.

FInal comments, don't use nested list, use numpy arrays for example `self.residuals[j,int(h / 3)]`.

AurelienRenou replied:
self.nb_dim done
iterate on self.nb_frames done 
replace self.sol by self.list_sol done

AurelienRenou replied:
self.name_of_variables as been change to name_of_variables

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -219,3 +234,21 @@ def animate(self):
+            residuals=self.residuals,
+            nb_iteration_diff=self.nb_iteration_diff,
+            nb_iteration_jac=self.nb_iteration_jac,
+        )
```
All the variables stored in output should not be attributes of the method.

replace   `self.residuals_xyz[i][j]` by  `residuals_xyz[i][j]`
Same for every other variables.

You should iterate on the `self.nb_frame` instead of the `self.sol`
Try to find a better one for `sol.fun`

To be more generic, you can create a new attribute `self.n_dim` related to the dimension of markers, we could have 2D motion capture so it would be:
`self.nb_dims * self.nb_markers` instead of `*3`

Anyway, your loop needs to be clarified.
Please start your outer loop with the number of markers.
There should be only one loop.

FInal comments, don't use nested list, use numpy arrays for example `self.residuals[j,int(h / 3)]`.

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -81,6 +82,17 @@ def __init__(
+        self.end = None
+        self.duration = None
+        # self.residuals_xyz = [[0]*self.nb_markers*3]*self.nb_frames
+        self.residuals_xyz = np.array([[0.0] * self.nb_markers * 3] * self.nb_frames, ndmin=2)
+        self.residuals = np.array([[0.0] * self.nb_markers] * self.nb_frames, ndmin=2)
+        self.nb_iteration_diff = []
+        self.nb_iteration_jac = []
+        self.output = None
```
see comment below, to be removed except for self.output which should be an empty dict.

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -81,6 +82,17 @@ def __init__(
+        self.end = None
+        self.duration = None
+        # self.residuals_xyz = [[0]*self.nb_markers*3]*self.nb_frames
+        self.residuals_xyz = np.array([[0.0] * self.nb_markers * 3] * self.nb_frames, ndmin=2)
+        self.residuals = np.array([[0.0] * self.nb_markers] * self.nb_frames, ndmin=2)
+        self.nb_iteration_diff = []
+        self.nb_iteration_jac = []
+        self.output = None
```
see comment below, to be removed except for self.output which should be an empty dict.

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -81,6 +82,17 @@ def __init__(
         self.q = np.zeros((self.nb_q, self.nb_frames))
         self.bounds = get_range_q(self.biorbd_model)
 
+        self.sol = []
```
to be removed

AurelienRenou replied:
I don't understand why

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -81,6 +82,17 @@ def __init__(
         self.q = np.zeros((self.nb_q, self.nb_frames))
         self.bounds = get_range_q(self.biorbd_model)
 
+        self.sol = []
```
to be removed

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -81,6 +82,17 @@ def __init__(
         self.bounds = get_range_q(self.biorbd_model)
 
+        self.sol = []
+        self.start = None
+        self.end = None
+        self.duration = None
```
same

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -81,6 +82,17 @@ def __init__(
         self.bounds = get_range_q(self.biorbd_model)
 
+        self.sol = []
+        self.start = None
+        self.end = None
+        self.duration = None
```
same

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -185,6 +197,7 @@ def solve(self, method: str = "lm", full: bool = False):
                         Usually the most efficient method for small unconstrained problems.
 
         """
+        self.start = time.time()
```
remove time

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -185,6 +197,7 @@ def solve(self, method: str = "lm", full: bool = False):
                         Usually the most efficient method for small unconstrained problems.
 
         """
+        self.start = time.time()
```
remove time

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -209,7 +221,10 @@ def solve(self, method: str = "lm", full: bool = False):
                 self.q[:, ii] = sol.x
+                self.sol.append(sol)
         print("Inverse Kinematics done for all frames")
+        self.end = time.time()
+        self.duration = self.end - self.start
```
say bye bye to these lines

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -209,7 +221,10 @@ def solve(self, method: str = "lm", full: bool = False):
                 self.q[:, ii] = sol.x
+                self.sol.append(sol)
         print("Inverse Kinematics done for all frames")
+        self.end = time.time()
+        self.duration = self.end - self.start
```
say bye bye to these lines

Ipuch reviewed (commented):
On its way

AurelienRenou committed:
review

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -236,19 +226,20 @@ def animate(self):
+            nb_iteration_diff.append(self.list_sol[i].nfev)
+            nb_iteration_jac.append(self.list_sol[i].njev)
+            residuals_xyz[i] = self.list_sol[i].fun
+            for h in range(0, self.nb_dim * self.nb_markers, self.nb_dim):
+                residuals[i][int(h / self.nb_dim)] = np.linalg.norm(self.list_sol[i].fun[h: h + self.nb_dim])
```
I'm sure this loop is not useful,
Plus, I dislike the form `residuals[i][int(h / self.nb_dim)]`, I would recommend `residuals[i, int(h / self.nb_dim)]`
maybe you should explain to me what you're doing.
I feel it is not the right thing.

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -236,19 +226,20 @@ def animate(self):
-        for j, value in enumerate(self.residuals_xyz):
-            for h in range(0, 3 * self.nb_markers, 3):
-                self.residuals[j][int(h / 3)] = np.linalg.norm(value[h : h + 3])
+        residuals_xyz = np.array([[0.0] * self.nb_markers * self.nb_dim] * self.nb_frames, ndmin=2)
+        residuals = np.array([[0.0] * self.nb_markers] * self.nb_frames, ndmin=2)
+        nb_iteration_diff = []
+        nb_iteration_jac = []
```
Use only arrays

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -82,16 +81,10 @@ def __init__(
+        self.list_sol = []
+
+        self.output = dict()
+        self.nb_dim = self.xp_markers.shape[0]
```
This should be on line 78.
Is the docstring updated?

AurelienRenou replied:
done

Ipuch reviewed (commented):
ça s'en vient

AurelienRenou committed:
update docstring, clear sol function and add max_marker, message, status and success list

AurelienRenou committed:
blacked

AurelienRenou committed:
reviewed by ipuch

AurelienRenou committed:
Merge remote-tracking branch 's2mLab/main' into output

handling conflicts

Ipuch commented:
Legit
